Plutarch/BitString.hs

```diff
@@ -0,0 +1,147 @@
+module Plutarch.BitString (
+  -- * Type
+  PBitString (..),
+
+  -- * Functions
+  preadBit,
+  psetBits,
+  pclearBits,
+  pshift,
+  protate,
+  pcountSetBits,
+  pfindFirstSetBit,
+  pfindFirstSetBit',
+) where
+
+import GHC.Generics (Generic)
+import Plutarch.Bool (PBool, PEq, POrd, PPartialOrd, pif, (#<))
+import Plutarch.Builtin (PBuiltinList)
+import Plutarch.ByteString (PByteString)
+import Plutarch.Integer (PInteger)
+import Plutarch.Internal (
+  S,
+  Term,
+  phoistAcyclic,
+  plet,
+  punsafeBuiltin,
+  (#),
+  (:-->),
+ )
+import Plutarch.Internal.Newtype (PlutusTypeNewtype)
+import Plutarch.Internal.Other (pto)
+import Plutarch.Internal.PLam (plam)
+import Plutarch.Internal.PlutusType (
+  DerivePlutusType (DPTStrat),
+  PlutusType,
+  pcon,
+ )
+import Plutarch.Lift (pconstant)
+import Plutarch.Maybe (PMaybe (PJust, PNothing))
+import PlutusCore qualified as PLC
+
+{- | A wrapper around 'PByteString' for CIP-122 and CIP-123 bitwise operations.
+
+= Note
+
+This type exists because /bit/ and /byte/ indexes work in different
+directions. To avoid confusing behaviour, we require an explicit wrapping of
+'PByteString's to use bitwise functionality: this way, it's clear where which
+scheme applies.
+
+@since WIP
+-}
+newtype PBitString (s :: S) = PBitString (Term s PByteString)
+  deriving stock
+    ( -- | @since WIP
+      Generic
+    )
+  deriving anyclass
+    ( -- | @since WIP
+      PlutusType
+    , -- | @since WIP
+      PEq
+    , -- | @since WIP
+      PPartialOrd
+    , -- | @since WIP
+      POrd
+    )
+
+-- | @since WIP
+instance DerivePlutusType PBitString where
+  type DPTStrat _ = PlutusTypeNewtype
+
+{- | Bit access operation, as defined in [CIP-122](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0122#readbit).
+
+@since WIP
+-}
+preadBit :: forall (s :: S). Term s (PBitString :--> PInteger :--> PBool)
+preadBit = plam $ \bs i -> punsafeBuiltin PLC.ReadBit # pto bs # i
```

Comment:
We don't need `plam` here and every places below. `punsafeBuiltin` will take care of it. This is actually less efficient